### ISSUE TITLE
[Bug][ROCm/gfx950] GLM-5.3-Flash 16K-chunk prefill: GPU memory-access fault when indexer triton kernels JIT during lazy CUDA-graph capture — clean under --enforce-eager

### ISSUE THREAD

**Issue by mustafayildirim** [vllm-project/vllm#57227]
### Environment

- vLLM: `v0.28.1rc1.dev681+ge7edf17ce` (V1 engine), aiter @ `2693202`, triton 3.7.0
- Hardware: 8× AMD MI350X (gfx950), TP8 + expert-parallel
- Model: GLM-5.3-Flash fp8 (MLA + KDA sparse attention indexer → uses aiter `batched_gemm_a16wfp4`, `_gluon_fp8_mqa_logits`, `_cp_gather_indexer_quant_cache_gfx950_kernel`)
- Config: `--max-num-batched-tokens 16384 --max-model-len 524288 --enable-expert-parallel --distributed-executor-backend mp`, default `cudagraph_mode=FULL_AND_PIECEWISE`, `VLLM_USE_BREAKABLE_CUDAGRAPH=1`

### Symptom

The **first** 16K-token chunk prefill crashes the engine with `Memory access fault` on all 8 ranks (e.g. `Memory access fault by GPU node-4 (Agent handle: 0x59ba0110) on address 0x7f733a000000. Reason: Unknown.`). Decode-only traffic and prefill chunks ≤8192 tokens are stable. Production impact: crash-looping pods on the first long-prompt request.

### Isolation: it's the graph capture, not the kernels

Differential with an identical request repro across cudagraph modes:

| Cudagraph mode | First 16K prefill |
|---|---|
| FULL | fault ×8 ranks |
| PIECEWISE | fault ×8 |
| FULL_AND_PIECEWISE | fault ×8 |
| **`--enforce-eager`** | **clean (0 faults, full repro suite)** |

At the moment of the fault, every kernel in the indexer chain is JIT-compiling *during inference* (boot warmup only exercised decode shapes; the 16K-prefill shapes are first seen at request time, when the 16384 cudagraph bucket is lazily captured):

```
WARNING [jit_monitor.py:141] Triton kernel JIT compilation during inference: BuildPrefillChunkMetadataKernel.kernel ...
WARNING [jit_monitor.py:141] Triton kernel JIT compilation during inference: _cp_gather_indexer_quant_cache_gfx950_kernel ...
WARNING [jit_monitor.py:141] Triton kernel JIT compilation during inference: _gluon_fp8_mqa_logits_kernel ...
WARNING [jit_monitor.py:141] Triton kernel JIT compilation during inference: _expand_pools_and_append_tail_kernel ...
WARNING [jit_monitor.py:141] Triton kernel JIT compilation during inference: _batched_gemm_a16wfp4_kernel ...
Memory access fault by GPU node-4 ...
```

### What we ruled out (in-server A/B, one full boot + repro per leg)

- Skipping `_gluon_fp8_mqa_logits` (correct-shape −inf logits substituted) → fault persists
- Skipping `_cp_gather_indexer_quant_cache_gfx950` (zero-fill) → fault persists
- Skipping `batched_gemm_a16wfp4` → **fault gone** (kernel execution required)
- Cloning **all** bmm tensor args (x, w, w_scales, y) to fresh contiguous memory in-server → fault persists (addresses innocent)
- `AMD_SERIALIZE_KERNEL=3` on/off → no effect
- Standalone replay with the **exact tensors dumped from the faulting call** (real strides/scales layout), M-sweep 4096→16000, workspace-pool carving with offset sweeps, two-stream interleave, plain `torch.cuda.CUDAGraph` capture, and fresh-variant-JIT-inside-plain-capture → **all clean**

So: the kernel is individually correct and capture-safe in isolation, eager execution is correct, but vLLM's lazy capture of the 16384 bucket (with fresh triton JIT in the same window) faults on gfx950. We could not reproduce it outside vLLM's capture machinery.

### Current workaround

`--enforce-eager` (fully stable at 16384) — at a significant decode-throughput cost. Chunk 8192 also avoids the fault but sacrifices prefill chunk size.

### Possible directions

1. Cover the 16K-prefill shapes in boot warmup *before* the lazy capture (so JIT and first launches happen eagerly, then capture) — if the fault is specifically first-launch-during-capture, this may be sufficient and cheap.
2. Investigate whether triton-on-ROCm first-launch/module-load inside a lazily-captured graph is a known-unsafe pattern (cc ROCm/aiter#5512 — same fault, tracked there with the full bisect matrix and tensor dumps).

Happy to provide the in-pod pytest harness (~6s per leg), tensor dumps, and boot logs.

**Comment by mustafayildirim** [comment]
## Follow-up: prewarming is NOT sufficient to fix this (tested)

We tested the "extend warmup to cover the prefill shapes" direction end-to-end on the failing setup (MI350X TP8, aiter @ `2693202`, `max_num_batched_tokens=16384`, cudagraphs on):

**Experiment:** patched `GPUModelRunner.capture_model` (vllm `v1/worker/gpu/model_runner.py`) to run one **eager** `_dummy_run(max_num_tokens=16384)` before any cudagraph capture.

**Result:**
- The eager 16384-token dummy prefill completed cleanly on all 8 ranks at boot (JIT'd the fresh-prefill specializations of `_batched_gemm_a16wfp4`, `_cp_gather_indexer_quant_cache_gfx950`, `_gluon_fp8_mqa_logits` outside capture).
- The first real chunked prefill **still faulted** (8× Memory access fault) **and still JIT-compiled the same kernels during inference** — because the faulting call is a *chunked* prefill with a computed prefix (`batched_gemm_a16wfp4` at `M=16000`, i.e. chunk minus prefix), a **different triton specialization** than the fresh 16384-token prefill the dummy run exercised (different chunk metadata, cu_seqlens, kv lengths, gather layouts).

**Conclusion:** the triton JIT specialization space for the chunked-indexer path is effectively unbounded (depends on prefix length, chunk boundaries, kv layout), so a deterministic prewarm cannot cover it. "Extend warmup" is not a viable fix for this bug.

Additional data point from earlier in the investigation: during the faulting call, `BreakableCUDAGraphCapture.current()` is `None` (the bmm executes in an eager segment between captured pieces, not inside a capture) — yet the fault only occurs with cudagraphs enabled. This suggests an interaction between the eager bmm segment and the surrounding captured graph pieces on gfx950, which likely needs attention from those familiar with the breakable-cudagraph machinery.

Current proven workaround remains `--enforce-eager`.

**Comment by simondanielsson** [comment]
Thanks @mustafayildirim!

I was able to reproduce this error using a 500k prompt in bench serve (see command below)
- `--max-num-batched-tokens 16384`: MAF
- `--max-num-batched-tokens 8192`: MAF
- `--max-num-batched-tokens 16384 --enforce-eager`: OK

using this base config. Both EP on and EP off triggers the MAF.
```bash
export VLLM_ROCM_USE_AITER=1
export VLLM_USE_BREAKABLE_CUDAGRAPH=1
vllm serve zai-org/GLM-5.3-Flash --tensor-parallel-size 8 --max-num-seqs 512 --gpu-memory-utilization 0.8 --attention-backend ROCM_AITER_MLA_SPARSE
```
Workload:
```
docker exec glm vllm bench serve \
  --backend vllm --model zai-org/GLM-5.3-Flash  \
  --dataset_name random --random_input_len 500000 --random_output_len 256 \
  --max_concurrency 4 --num_prompts 16 --num_warmups 8 \
  --host localhost --port 8000 --ignore_eos --seed 5679
```
=> looks like not specific to 16k max num batched tokens.

I was **not** able to trigger it with a 32k prompt and `--max-num-batched-tokens 16384` though:
```
docker exec glm vllm bench serve \
  --backend vllm --model zai-org/GLM-5.3-Flash  \
  --dataset_name random --random_input_len 32000 --random_output_len 256 \
  --max_concurrency 4 --num_prompts 16 --num_warmups 8 \
  --host localhost --port 8000 --ignore_eos --seed 5679
```

Image: vllm/vllm-openai-rocm:nightly-af1c01499b289be555c475669ba50a88e96d846e + https://github.com/vllm-project/vllm/pull/57192 + https://github.com/vllm-project/vllm/pull/56604

**Comment by mustafayildirim** [comment]
## The official MI350X recipe config hits the same fault

We verified the official serving recipe from recipes.vllm.ai (zai-org/GLM-5.3-Flash, MI350X, single-node TEP) against this bug.

**Recipe command** (docker `vllm/vllm-openai-rocm:nightly`, tested on our MI350X TP8 pod with vLLM `v0.28.1rc1.dev681+ge7edf17ce`, aiter `2693202`):

```
--enable-expert-parallel --tensor-parallel-size 8 --max-num-seqs 512 \
--attention-backend ROCM_AITER_MLA_SPARSE \
--tool-call-parser glm47 --enable-auto-tool-choice --reasoning-parser glm45
```

(no `--max-num-batched-tokens`, no `--enforce-eager` — exactly as published)

**Results:**
- Boot: clean, Ready. Note the engine derives **`max_num_batched_tokens=16384` by default** for this model — the recipe does not actually run at a smaller chunk size.
- First 20K-token prompt (chunked prefill, chunk 2 = M=16000 with prefix): **Memory access fault ×8 ranks — identical crash** to the configurations reported above.

**Why the recipe's validation likely missed this:** the published benchmark is `--random-input-len 8192` — no request ever produces a prefill chunk larger than 8192, and the fault requires a chunk that fills the 16384 budget (a chunk with a computed prefix, i.e. M between ~8192 and 16384 on a later chunk of a longer prompt). Single-chunk 16K prefills and all decode traffic are unaffected.

**Implication:** any user following the official MI350X recipe and serving prompts longer than ~16K tokens will crash-loop on the first long request, exactly as our production deployment did.

**Suggested interim documentation fix:** add `--max-num-batched-tokens 8192` to the MI350X recipe until the underlying capture/JIT fault is fixed (this caps all prefill chunks at 8192 and is verified stable), or note the limitation explicitly.

**Comment by mustafayildirim** [comment]
## ROOT CAUSE FOUND + working fix: `VLLM_USE_BREAKABLE_CUDAGRAPH=0`

### Root cause

The fault is triggered by the **breakable cudagraph machinery** (`VLLM_USE_BREAKABLE_CUDAGRAPH=1`), not by the aiter kernels or any memory layout.

Differential (identical request repro, MI350X TP8):

| Config | First 16K-chunk prefill |
|---|---|
| breakable=1 (default in our env), cudagraph_mode FULL / PIECEWISE / FULL_AND_PIECEWISE | **fault ×8 ranks** |
| breakable=1, `--enforce-eager` | clean |
| **breakable=0, cudagraphs on (FULL_AND_PIECEWISE, 57 FULL graphs captured)** | **clean — 0 faults, repro ×2** |

### Working fix

```
VLLM_USE_BREAKABLE_CUDAGRAPH=0
```

- Validated on vLLM `0.29.1rc1.dev187+gaf1c01499` (nightly, 2026-09-16) with aiter `2693202`: full repro suite ×2 passes with zero faults, `max_num_batched_tokens=16384` honored, FULL+PIECEWISE cudagraphs active, decode graphs intact.
- On older builds without torch.compile support for this model (`0.28.1rc1.dev681+ge7edf17c`), breakable=0 cannot be used — the engine raises `piecewise CUDA graphs unavailable, model is not torch-compiled and breakable CUDA graph is off` (that build has no piecewise mechanism other than breakable graphs). The fix requires a torch.compile-enabled build.

### Mechanism notes for maintainers

- The faulting kernel (`aiter batched_gemm_a16wfp4` at M=16000, chunk 2 of a chunked prefill) executes with `BreakableCUDAGraphCapture.current() is None` — i.e. in an **eager segment between breakable-capture pieces** — yet faults only when the breakable machinery is active. Kernel serialization on/off, all tensor addresses (cloned in-server), M-threshold, and every standalone replay (including plain `torch.cuda.CUDAGraph` capture and fresh-JIT-in-capture) are clean. This points at the eager-segment-on-capture-stream execution pattern in `vllm/compilation/breakable_cudagraph.py` (ending a stream capture, launching a triton kernel on the capture stream, resuming capture) interacting badly with gfx950 — possibly the triton launch path doing module-load/scratch-alloc between capture end and resume.
- Triton JIT-during-inference still occurs with breakable=0 (7 warmup warnings at first prefill) and is harmless there — JIT alone is not the trigger.

### Repro summary

Serve GLM-5.3-Flash fp8 on 8×MI350X, TP8, `--max-num-batched-tokens 16384 --max-model-len 524288`, breakable cudagraph on (default), then send a prompt >16K tokens (chunk 2 of the chunked prefill, M=16000 with computed prefix). Memory access fault on all ranks, engine dies, pod crash-loops. Repro scripts and full bisect matrix in earlier comments.

**Comment by simondanielsson** [comment]
@mustafayildirim Please try this fix: https://github.com/vllm-project/vllm/pull/57317 (in addition to https://github.com/vllm-project/vllm/pull/57192 and https://github.com/vllm-project/vllm/pull/56604). 

With this I was able to run 500k prompts without issues:

| input_len | prompts | concurrency | breakable cudagraph | unpatched | patched |
|---|---|---|---|---|---|
| 67000 | 1 | 1 | 0 | MAF x8 | clean |
| 67000 | 1 | 1 | 1 | MAF x8 | clean |
| 500000 | 1 | 1 | 0 | MAF x8 | clean |
| 500000 | 1 | 1 | 1 | MAF x8 | clean |
| 500000 | 4 | 2 | 0 | MAF x8 | clean |

using

```sshell
export VLLM_ROCM_USE_AITER=1
export VLLM_USE_BREAKABLE_CUDAGRAPH=1
vllm serve zai-org/GLM-5.3-Flash --tensor-parallel-size 8 --max-num-seqs 512 --gpu-memory-utilization 0.8 --attention-backend ROCM_AITER_MLA_SPARSE --max-num-batched-tokens 16384
```

**Comment by mustafayildirim** [comment]
Confirmed: with #57317 (+ #57425 for a separate nightly boot blocker, + #56604, + #57192) applied on nightly `0bfc7a15d` on 8×MI350X TP8, the engine runs with `VLLM_USE_BREAKABLE_CUDAGRAPH=1` and passes chunked-prefill repro + long-context ladder up to 524,212 tokens with zero memory access faults. #57317 is the root cause — breakable cudagraphs were only an amplifier, not the bug.
